### DISCUSSION TITLE
citekeys

### DISCUSSION THREAD

**Discussion by Halem998** [papis/papis#595] [Q&A] [answered]
Back in Zotero (which I'm trying to ditch) there are plugins like `zotfile` and `betterbibtex`.

- with zotfile I can define the citekey, e.g. by defining it as {%b} etc etc.

- with betterbibtex I can refresh citekeys ( let's say I've corrected some info that would change the citekey)

How do I do these things in papis?

**Comment by alexfikl** [comment, marked as answer]
I'm not very familiar with Zotero or its plugins, so not quite sure if this is a good replacement. papis uses `ref` when referring to citekeys, and those usually get autogenerated when you add a new document using the format
https://papis.readthedocs.io/en/latest/configuration.html#config-settings-ref-format

If you want to update the reference for an existing document, you could use something like
```
papis update --set ref 'MyNewRef' <query>
papis update --set ref '{doc[author_list][0][family]}{doc[year]}' <query>
```
or whatever else you want to put in there for new reference. From a quick look at `betterbibtex`, it seems like they have more advanced features to format the reference, so not sure papis covers everything there at the moment.

Does that help?

**Reply by Halem998** [reply]
Yeah, I can live with just bibtex. The CLI nature of papis is so great I just wanna get rid of zotero.

Also I would love to have a mechanism that when changing the citekey, to have the folder name and also the document name changed accordingly.

**Comment by Halem998** [comment]
I did some search and I think I found what I was looking for! Thanks @alexfikl 

In the config, it's `add-folder-name=`, `add-file-name` and `ref-format =`.

So now I can just set this all to identical values ( I believe this is better for searchability?)

But I'm not quite familiar with the syntax used here. 

How do I make `{doc[author_list][0][family]}{doc[year]}`
to appear with underscores, like so:  `author_title_year` ?

Also, is there a way I can let papis generate a unique title name? 
Take for example "A Symbolic Analysis of Relay and Switching Circuits" -- that's a pretty long name. 
In Zotero it was convenient, it would shorten the title to, say, "Symbolic", also dismissing "A", thus taking the first meaningful word in the title or something along that line ...
So this would give me `Shannon_Symbolic_1937`
But I guess most likely I have to learn some coding and write this myself

**Reply by alexfikl** [reply]
> But I'm not quite familiar with the syntax used here.
> How do I make `{doc[author_list][0][family]}{doc[year]}`
> to appear with underscores, like so: author_title_year ?

This just uses the standard python formatting syntax, which you get with `ref_format.format(doc=doc)`. You can check out more advanced usage in the Python documentation at
https://docs.python.org/3/library/string.html#formatstrings
To get the underscores, you could just use `{doc[author_list][0][family]}_{doc[year]}`. You can't really do fancier things here, since python doesn't allow it :(

papis also supports formatting strings based on the Jinja templating language, which you can set with
https://papis.readthedocs.io/en/latest/configuration.html#config-settings-formater
That may give you fancier formatting options, but it's a bit more work to set up all the different settings, since the defaults are all for the python formatter.

>  But I guess most likely I have to learn some coding and write this myself

For custom usage, that's the best choice! The core papis tries to be very flexible, so you can add simple plugins to do what you need.

**Comment by jghauser** [comment]
I have a setup that's using the jinja2 formatter. I keep the relevant configuration in [config.py](https://papis.readthedocs.io/en/latest/configuration.html#python-configuration-file) because it's easier to get escaping/quoting etc. right. Here's what I got:

```python
import papis.config

options = {
    "settings": {
        "ref-format": '{% if doc["author_list"] %}'
        + '{{(doc["author_list"][0].get("surname","")[0:10] or doc["author_list"][0].get("family","")[0:10])}} '
        + '{% elif doc["author"] %}'
        + '{{doc["author"].split()[0][0:10]}} '
        + '{% elif doc["editor"] %}'
        + '{{doc["editor"].split()[0][0:10]}} '
        + "{% endif %}"
        + '{% if doc["year"] %}'
        + '{{doc["year"]}} '
        + "{% endif %}"
        + '{% for item in doc["title"].split() %}'
        + "{% if loop.index <= 3 %}"
        + "{{ item[0:3]|title }} "
        + "{% endif %}"
        + "{% endfor %}",
    }
}
papis.config.register_default_settings(options)
```

This gets you references like: `Abelson1996StrAndInt`, So, name of author, year, and first three characters of the first three words (which gives me a nice trade-off of concision and information). I'm just adding it here in case you want to see a jinja2 example. Note that if you set `formater = jinja2` in your `config` file, you will need to define jinja2 versions for the other config options you mentioned. Happy to provide more info if you're interested!

**Reply by Halem998** [reply]
this looks great! I’m still in a hacky stage when I copy paste code I dont understand :) But Database related stuff is a rabbit hole I can’t stop following. Thanks again, Iet’s see how far I’ll get

**Comment by avonmoll** [comment]
One of the things that Jinja2 can't do (unless I'm wrong, then please let me know!!) is filter out "stop words" like "a", "the", "on" at the beginning of titles. My desired refformat is `<first author last name><year><first non stop-word of title>`. What I've done may be _way_ overkill, but I have a post-processing script which updates the `ref`s of my whole library and also makes the folder/file names conform to my desired format.

Regarding stop words, I import the `nltk` package and I have a function in my script which will get the first `K` stopwords:

```python
import nltk
from nltk.corpus import stopwords
sw = set(stopwords.words("english"))

def getKthNonstopword(d, K):
    title_words = d["title"].lower().split(" ")
    i = -1 
    k = 0
    while i < len(title_words) - 1 and k < K:
        i += 1
        if title_words[i] not in sw:
            k += 1
    return title_words[i]
```

The reason I get up to `K` words is that sometimes the `ref` will not be unique under my spec (e.g., an author publishes two papers in the same year with the same first word in the title)

Then I have a function to build the desired `ref`:

```python
def ref(d):
    r = getFirstAuthorName(d).lower().replace(" ", "")
    r = unicodedata.normalize("NFKD", r).encode("ascii", "ignore").decode("utf8")
    if "year" in d:
        r += str(d["year"])
    title_word = getKthNonstopword(d, 1)
    r += title_word
    r = r.replace(".", "")
    r = r.replace("{", "")
    r = r.replace("}", "")
    r = r.replace(",", "")
    r = r.replace("/", "")
    r = r.replace("\\", "")
    r = r.replace("`", "")
    r = r.replace("\'", "")
    return r
```

Then my script loops through all the documents in my library and updates the `refs` (and folder and file names, which I did not show, accordingly):

```python
refs = []
directories = os.listdir(directory)
for i, docfolder in enumerate(sorted(directories)):
    # skip files
    if not os.path.isdir(os.path.join(directory, docfolder)):
        continue
    contents = os.listdir(os.path.join(directory, docfolder))
    if "info.yaml" not in contents:
        continue
    with open(os.path.join(directory, docfolder, "info.yaml")) as fd:
        doc = yaml.safe_load(fd)
# ...
    r = ref(doc)  # compute new ref field
    
    # write changed doc to the info.yaml file
    with open(os.path.join(directory, Fname, "info.yaml"), "w") as fd:
        yaml.dump(doc, fd, default_flow_style=False, allow_unicode=True)
```

**Reply by jghauser** [reply]
Wow, That is impressive :exploding_head:! Regarding filtering out 'stop words' with jinja2, I think that might actually be possible. At least I don't as such see why you couldn't filter them out with some 'if' statement(s). I haven't checked though, and it might definitely get a bit unwieldy.
